chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,245 @@
+import re
+import numpy as np
+from astropy.table import Table, vstack
+from kadi import events
+from Ska.Numpy import interpolate
+from Quaternion import Quat
+import Ska.quatutil
+from mica.archive import asp_l1
+import mica.starcheck
+from Ska.engarchive import fetch
+from Chandra.Time import DateTime
+import agasc
+
+
+R2A = 206264.81  # Convert from radians to arcsec
+
+
+def quat_vtransform(qs):
+    """
+    Transform a Nx4 matrix of quaternions into a Nx3x3 transform matrix
+    :returns: Nx3x3 transform matrix
+    :rtype: numpy array
+    """
+    x, y, z, w = qs[:, 0], qs[:, 1], qs[:, 2], qs[:, 3]
+    xx2 = x * x * 2.
+    yy2 = y * y * 2.
+    zz2 = z * z * 2.
+    xy2 = x * y * 2.
+    wz2 = w * z * 2.
+    zx2 = z * x * 2.
+    wy2 = w * y * 2.
+    yz2 = y * z * 2.
+    wx2 = w * x * 2.
+
+    t = np.empty((len(qs), 3, 3), float)
+    t[:, 0, 0] = 1. - yy2 - zz2
+    t[:, 0, 1] = xy2 - wz2
+    t[:, 0, 2] = zx2 + wy2
+    t[:, 1, 0] = xy2 + wz2
+    t[:, 1, 1] = 1. - xx2 - zz2
+    t[:, 1, 2] = yz2 - wx2
+    t[:, 2, 0] = zx2 - wy2
+    t[:, 2, 1] = yz2 + wx2
+    t[:, 2, 2] = 1. - xx2 - yy2
+
+    return t
+
+
+def quat_vmult(q1, q2):
+    # just assume q1 and q2 have same length
+    mult = np.zeros((len(q1), 4))
+    mult[:,0] =  q1[:,3] * q2[:,0] - q1[:,2] * q2[:,1] + q1[:,1] * q2[:,2] + q1[:,0] * q2[:,3]
+    mult[:,1] =  q1[:,2] * q2[:,0] + q1[:,3] * q2[:,1] - q1[:,0] * q2[:,2] + q1[:,1] * q2[:,3]
+    mult[:,2] = -q1[:,1] * q2[:,0] + q1[:,0] * q2[:,1] + q1[:,3] * q2[:,2] + q1[:,2] * q2[:,3]
+    mult[:,3] = -q1[:,0] * q2[:,0] - q1[:,1] * q2[:,1] - q1[:,2] * q2[:,2] + q1[:,3] * q2[:,3]
+    return mult
+
+
+class CentroidResiduals(object):
+    centroid_source = None
+    att_source = None
+    ra = None
+    dec = None
+
+    def set_centroids(self, start, stop, source, slot, alg=8):
+        """
+        Get centroids from ``source`` between ``start`` and ``stop``.
+
+        One could also just set yags, yag_times, zags, zag_times
+        attributes directly.
+
+        :param start: start time
+        :param stop: stop time
+        :param source: 'ground' | 'obc'
+        :param slot: ACA slot
+        """
+        self.centroid_source = source
+        # Get centroids from Ska eng archive or mica L1 archive
+        # Might want to add filtering for track status here too
+        # Also need to include something for time offsets
```

Comment:
@malgosias Where can I find that work of yours that showed some centroid time offsets?  I am forgetting the details.